staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go

```diff
@@ -0,0 +1,341 @@
+			return nil, fmt.Errorf("request to convert CR from an invalid group/version: %s", fromGVK.String())
+		}
+		if fromGVK == toGVK {
+			// No conversion is required
+			return in, nil
+		}
+	}
+
+	request := createConversionReview(in, toGVK.GroupVersion().String())
+	response := &v1beta1.ConversionReview{}
+	// TODO: Figure out if adding one second timeout make sense here.
+	ctx := context.TODO()
+	r := c.restClient.Post().Context(ctx).Body(request).Do()
+	if err := r.Into(response); err != nil {
+		// TODO: Return a webhook specific error to be able to convert it to meta.Status
+		return nil, fmt.Errorf("calling to conversion webhook failed for %s: %v", c.name, err)
+	}
+
+	if response.Response == nil {
+		// TODO: Return a webhook specific error to be able to convert it to meta.Status
+		return nil, fmt.Errorf("conversion webhook response was absent for %s", c.name)
+	}
+
+	if response.Response.Result.Status != v1.StatusSuccess {
+		// TODO return status message as error
+		return nil, fmt.Errorf("conversion request failed for %v, Response: %v", in.GetObjectKind(), response)
+	}
+
+	if len(response.Response.ConvertedObjects) != len(request.Request.Objects) {
+		return nil, fmt.Errorf("expected %v converted objects, got %v", len(request.Request.Objects), len(response.Response.ConvertedObjects))
+	}
+
+	if isList {
+		convertedList := listObj.DeepCopy()
+		// Collection of items sent for conversion is different than list items
+		// because only items that needed conversion has been sent.
+		convertedIndex := 0
+		for i := 0; i < len(listObj.Items); i++ {
+			if listObj.Items[i].GetAPIVersion() == toGVK.GroupVersion().String() {
+				// This item has not been sent for conversion, skip it.
+				continue
+			}
+			converted, err := getRawExtensionObject(response.Response.ConvertedObjects[convertedIndex])
+			original := request.Request.Objects[convertedIndex].Object
+			convertedIndex++
+			if err != nil {
+				return nil, fmt.Errorf("invalid converted object at index %v: %v", convertedIndex, err)
+			}
+			if e, a := toGVK.GroupVersion(), converted.GetObjectKind().GroupVersionKind().GroupVersion(); e != a {
+				return nil, fmt.Errorf("invalid converted object at index %v: invalid groupVersion, e=%v, a=%v", convertedIndex, e, a)
+			}
+			if e, a := original.GetObjectKind().GroupVersionKind().Kind, converted.GetObjectKind().GroupVersionKind().Kind; e != a {
+				return nil, fmt.Errorf("invalid converted object at index %v: invalid kind, e=%v, a=%v", convertedIndex, e, a)
+			}
+			unstructConverted, ok := converted.(*unstructured.Unstructured)
+			if !ok {
+				// this should not happened
+				return nil, fmt.Errorf("CR conversion failed")
+			}
+			if err := validateConvertedObject(&listObj.Items[i], unstructConverted); err != nil {
+				return nil, fmt.Errorf("invalid converted object at index %v: %v", convertedIndex, err)
+			}
+			convertedList.Items[i] = *unstructConverted
+		}
+		convertedList.SetAPIVersion(toGVK.GroupVersion().String())
```

Comment:
if the input was a `v1` List containing `example.com/v1alpha1` Widget items, requesting conversion to `example.com/v1beta1`, wouldn't this result in an `example.com/v1alpha1` List as output?

Reply:
well, it depends what `target.KindForGroupVersionKinds` meant to do. For GroupVersion.target.KindForGroupVersionKinds (which I think the type of target we pass to conversion is GroupVersion right now) it will only replace kind with the kind of given GVK, so 

```
toGVK, ok := target.KindForGroupVersionKinds([]schema.GroupVersionKind{fromGVK})
```

target is (example.com, v1beta1), fromGVK is (example.com, v1, ListType), then the toGVK will be (example.com, v1beta1, List) and as we only use GroupVersion here, it will be the right APIVersion example.com/v1beta1.

How can any of these being affected by the list item's APIVersion?

Reply:
I cleared the code up to only have toGV moving forward from initial checks. We don't need anything else for conversion. `toGV.String()` is `APIVersion`.

Reply:
That was why I asked what types of lists we support passing to this conversion function. We have two types of lists that get passed around:
* generic `v1` List objects containing non-homogenous objects of any arbitrary GVK. If you were going to read a bunch of objects directly from storage into a list, it would be this type (because storage does not guarantee all stored objects are in the same GVK). For this type of list, using the list's GVK to compute to/from for the items is incorrect, and setting the requested target version as the list's GV is also incorrect.
* specific GVK list objects (like `example.com/v1beta1` `MyKindList`) containing only objects with the same GVK (`example.com/v1beta1` `MyKind`). For this type of list, using the GV of the List to compute to/from for the items is more correct, and setting the requested target version as the list's GV would also be correct.

The analogue for built-in types would be:

Generic list:
* in:
    ```
    kind: List
    apiVersion: v1
    items:
    - {"kind":"Deployment","apiVersion":"apps/v1beta1",...}
    - {"kind":"Deployment","apiVersion":"apps/v1beta2",...}
   ```
* converted to `extensions/v1beta1`
* out:
    ```
    kind: List
    apiVersion: v1
    items:
    - {"kind":"Deployment","apiVersion":"extensions/v1beta1",...}
    - {"kind":"Deployment","apiVersion":"extensions/v1beta1",...}
    ```

Specific list:
* in:
    ```
    kind: DeploymentList
    apiVersion: apps/v1beta1
    items:
    - {"kind":"Deployment","apiVersion":"apps/v1beta1",...}
    - {"kind":"Deployment","apiVersion":"apps/v1beta1",...}
    ```
* converted to `extensions/v1beta1`
* out:
    ```
    kind: DeploymentList
    apiVersion: extensions/v1beta1
    items:
    - {"kind":"Deployment","apiVersion":"extensions/v1beta1",...}
    - {"kind":"Deployment","apiVersion":"extensions/v1beta1",...}
    ```

Reply:
Talked offline. v1 List objects aren't passed here (and if they were, the validVersions[fromGVK] check would fail and prevent insanity).

CRD storage impl creates an unstructured list with the request's GV, populates it from storage, then calls conversion to convert the individual items.

Specific list:
* in:
    ```
    kind: MyKindList
    apiVersion: acme.com/v1
    items:
    - {"kind":"MyKind","apiVersion":"acme.com/v1",...}
    - {"kind":"MyKind","apiVersion":"acme.com/v1beta1",...}
    ```
* converted to `acme.com/v1`
* out:
    ```
    kind: MyKindList
    apiVersion: acme.com/v1
    items:
    - {"kind":"MyKind","apiVersion":"acme.com/v1",...}
    - {"kind":"MyKind","apiVersion":"acme.com/v1",...}
    ```

this differs from built-in resources which cannot contain non-homogenous kinds in a specifically typed list

Reply:
CRDs do not use generic list. The list always has the type of the request and it is different than standard type codepath. Decoder will put objects with different GVK in an unstructuredList. The unstructuredList is created by the request store object which will have the GVK of the request. Thus here we assume even if the input is the list, the GV is what we want.
We are protected against the generic list because we check for GVK to be valid at the start of the conversion and a generic v1 list will not pass that.